commands2/commandscheduler.py

```diff
@@ -0,0 +1,471 @@
+        using those requirements have been scheduled as interruptible. If this is the case, they will
+        be interrupted and the command will be scheduled.
+
+        :param commands: the commands to schedule.
+        """
+        if len(commands) > 1:
+            for command in commands:
+                self.schedule(command)
+            return
+
+        command = commands[0]
+
+        if command is None:
+            # DriverStation.reportWarning("CommandScheduler tried to schedule a null command!", True)
+            return
+
+        if self._inRunLoop:
+            self._toSchedule.add(command)
+            return
+
+        if command in self.getComposedCommands():
+            raise IllegalCommandUse(
+                "A command that is part of a CommandGroup cannot be independently scheduled"
+            )
+
+        if self._disabled:
+            return
+
+        if RobotState.isDisabled() and not command.runsWhenDisabled():
+            return
+
+        if self.isScheduled(command):
+            return
+
+        requirements = command.getRequirements()
+
+        if self._requirements.keys().isdisjoint(requirements):
+            self.initCommand(command, *requirements)
+        else:
+            for requirement in requirements:
+                requiringCommand = self.requiring(requirement)
+                if (
+                    requiringCommand is not None
+                    and requiringCommand.getInterruptionBehavior()
+                    == InterruptionBehavior.kCancelIncoming
+                ):
+                    return
+
+            for requirement in requirements:
+                requiringCommand = self.requiring(requirement)
+                if requiringCommand is not None:
+                    self.cancel(requiringCommand)
+
+            self.initCommand(command, *requirements)
+
+    def run(self):
+        """
+        Runs a single iteration of the scheduler. The execution occurs in the following order:
+
+        Subsystem periodic methods are called.
+
+        Button bindings are polled, and new commands are scheduled from them.
+
+        Currently-scheduled commands are executed.
+
+        End conditions are checked on currently-scheduled commands, and commands that are finished
+        have their end methods called and are removed.
+
+        Any subsystems not being used as requirements have their default methods started.
+        """
+        if self._disabled:
+            return
+        self._watchdog.reset()
+
+        for subsystem in self._subsystems:
+            subsystem.periodic()
+            if RobotBase.isSimulation():
+                subsystem.simulationPeriodic()
+            # self._watchdog.addEpoch()
+
+        loopCache = self._activeButtonLoop
+        loopCache.poll()
+        self._watchdog.addEpoch("buttons.run()")
+
+        self._inRunLoop = True
+
+        for command in self._scheduledCommands.copy():
```

Comment:
`_scheduledCommands` is a set, which doesn't have a stable iteration order. The Java version is a LinkedHashSet which does, and the C++ version uses a SmallSet which defaults to a vector which also does. Seems like this should also? (use a `dict[Command, None]`?)

Reply:
In Java, all of them are ordered so I moved more to dicts. It didn't affect the tests either way so thats good.